### DISCUSSION TITLE
Action List implementation

### DISCUSSION THREAD

**Discussion by michaelfung** [mojolicious/mojo#1867] [General]
I need to process many directives coming into a single endpoint. I use an action list inside a controller like below:
Is there room for improvement or better alternatives? Thanks.

```
# in a controller

my %action_list = (
    foo => sub($c, $p) {
      ...
    },
    bar => sub($c, $p) {
      ...
    },
    # ... and many more

    default => sub($c, $p) {
        $c->render(status => 400, text => 'invalid directive');
    }
);

sub parse($c) {
    my $m = $c->param('m') // 'default';
    my $p = $c->param('p') // 0;
    $m = 'default' unless defined $action_list{$m};
    $action_list{$m}->($c, $p);
}
```

**Comment by kiwiroy** [comment]
Can you render separate templates using `action`, or is there more going on in those subs? See [link check example]( https://github.com/mojolicious/minion/blob/f4bc5b2e98aa619b88cd1c6188649f281be1c49e/examples/linkcheck/lib/LinkCheck/Controller/Links.pm#L7)

Separate templates and methods
```perl
sub bar { return shift() ^ 2 }
sub foo { return shift() * -1 }

sub parse($c) {
  my $m      = $c->param('m') // 'default';
  my $p      = $c->param('p') // 0;
  my $method = {foo => \&foo, bar => \&bar}->{$m};

  return $c->render(action => $m, p => $method->($p)) if $method;
  return $c->render(action => 'default', status => 400);
}
```

foo.html.ep
```html
foo with p value <%= $p %>
```
There's also [shortcuts](https://docs.mojolicious.org/Mojolicious/Guides/Routing#Shortcuts) and [placeholder types](https://docs.mojolicious.org/Mojolicious/Guides/Routing#Placeholder-types), but that might not fit with your route specification.

**Reply by michaelfung** [reply]
There are quite some processing in the subs. Output is not of concern(just success or failure). Thanks.

**Reply by jhthorsen** [reply]
I would go in the same direction, but with something like this:

```
sub parse ($c) {
  return $c->render(status => 400) unless my $action = $c->can(sprintf '_action_%s', $c->param('m') // 'default');
  return $c->render(p => $c->$action($c->param('p') // 0));
}

sub _action_bar { ... }
sub _action_foo { ... }
```

**Reply by michaelfung** [reply]
Wow, neat! Thanks!
